loadFromHTMLWithData only ever calls back on the main queue, creating deadlocks.

kieransenior opened the pull request:
#### dbf953adf9a729d64779885c797c8f7645e1826d
<pre>
loadFromHTMLWithData only ever calls back on the main queue, creating deadlocks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244670">https://bugs.webkit.org/show_bug.cgi?id=244670</a>
&lt;rdar://99404547&gt;

Reviewed by NOBODY (OOPS!).

Accept a queue parameter in the NSAttributedString category static
methods allowing clients to ensure callbacks do not encounter
deadlocks in the event callers higher up the stack are utilizing
semaphores or spinning the runloop on the same queue.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.h:
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NSAttributedStringWebKitAdditions.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbf953adf9a729d64779885c797c8f7645e1826d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87898 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18608 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97062 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152115 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30377 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26389 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80002 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91805 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93512 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24507 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74595 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24495 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79460 "Found 1 new API test failure: TestWebKitAPI.NSAttributedStringWebKitAdditions.DirectoriesNotCreatedBackgroundQueue (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14429 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/31224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37331 "Passed tests") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33704 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.h:
```diff
@@ -72,6 +72,21 @@ typedef void (^NSAttributedStringCompletionHandler)(NSAttributedString * _Nullab
++ (void)loadFromHTMLWithRequest:(NSURLRequest *)request options:(NSDictionary<NSAttributedStringDocumentReadingOptionKey, id> *)options
+    queue:(dispatch_queue_t _Nullable)queue
+    completionHandler:(NSAttributedStringCompletionHandler)completionHandler
+    NS_SWIFT_NAME(loadFromHTML(request:options:queue:completionHandler:)) WK_SWIFT_ASYNC_NAME(fromHTML(request:options:queue:)) WK_API_AVAILABLE(macos(10.15), ios(13.0));
```
API availability should be WK_MAC_TBA and WK_IOS_TBA, not these long-shipped versions.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.h:
```diff
@@ -72,6 +72,21 @@ typedef void (^NSAttributedStringCompletionHandler)(NSAttributedString * _Nullab
+ @param request The request specifying the URL to load.
+ @param options Document attributes for interpreting the document contents.
+ NSTextSizeMultiplierDocumentOption and NSTimeoutDocumentOption are supported option keys.
+ @param queue The queue to call back on
```
Nit: Missing period.

(3 other instances below)

kieransenior replied:
Fixed, thanks 👍🏼

kieransenior committed:
loadFromHTMLWithData only ever calls back on the main queue, creating deadlocks.
https://bugs.webkit.org/show_bug.cgi?id=244670
<rdar://99404547>

Reviewed by NOBODY (OOPS!).

Accept a queue parameter in the NSAttributedString category static
methods allowing clients to ensure callbacks do not encounter
deadlocks in the event callers higher up the stack are utilizing
semaphores or spinning the runloop on the same queue.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.h:
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NSAttributedStringWebKitAdditions.mm:
